scio-extra/src/main/scala/com/spotify/scio/extra/Sparkey.scala

```diff
@@ -0,0 +1,173 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.extra
+
+import java.io._
+import java.net.URI
+import java.nio.ByteBuffer
+import java.nio.file.{Paths, Files => JFiles}
+import java.util.UUID
+
+import com.google.common.base.Charsets
+import com.google.common.hash.Hashing
+import com.spotify.scio.ScioContext
+import com.spotify.scio.util.{CallSites, ScioUtil}
+import com.spotify.scio.values.{SCollection, SideInput}
+import com.spotify.sparkey.{CompressionType, SparkeyReader, Sparkey => JSparkey}
+import org.apache.beam.sdk.options.PipelineOptionsFactory
+import org.apache.beam.sdk.transforms.{DoFn, View}
+import org.apache.beam.sdk.util.GcsUtil.GcsUtilFactory
+import org.apache.beam.sdk.util.gcsfs.GcsPath
+import org.apache.beam.sdk.values.PCollectionView
+
+import scala.collection.JavaConverters._
+
+object Sparkey {
+
+  /**
+   * Represents the base URI for a Sparkey index and log file, either on the local file
+   * system or on GCS. For GCS, the URI should be in the form
+   * 'gs://<bucket>/<path>/<sparkey-prefix>'. For local files, the URI should be in the form
+   * '/<path>/<sparkey-prefix>'. Note that the URI must not be a folder or GCS bucket as the URI is
+   * a base path representing two files - <sparkey-prefix>.spi and <sparkey-prefix>.spl.
+   */
+  trait SparkeyUri {
+    val basePath: String
+    def getReader(): SparkeyReader
+    override def toString(): String = basePath
+  }
+
+  object SparkeyUri {
+    def apply(path: String): SparkeyUri =
+      if (ScioUtil.isGcsUri(new URI(path))) new GcsSparkeyUri(path) else new LocalSparkeyUri(path)
+  }
+
+  implicit class SparkeyScioContext(val self: ScioContext) extends AnyVal {
+    def sparkeySideInput(url: SparkeyUri): SideInput[SparkeyReader] = {
+      val view = self.parallelize(Seq(url)).applyInternal(View.asSingleton())
+      new SparkeySideInput(view)
+    }
+  }
+
+  implicit class SparkeyPairSCollection(val self: SCollection[(String, String)]) {
+    /**
+     * Write the contents of this SCollection as a Sparkey file, either locally or on GCS.
+     *
+     * @return A singleton SCollection containing the [[SparkeyUri]] of the saved files.
+     */
+    def asSparkey(uri: SparkeyUri): SCollection[SparkeyUri] = self.transform { in =>
+      in.groupBy(_ => ())
+        .map { case (_, iter) =>
+          val writer = new SparkeyWriter(uri)
+          val it = iter.iterator
+          while (it.hasNext) {
+            val kv = it.next()
+            writer.put(kv._1.toString, kv._2.toString)
+          }
+          writer.close()
+          uri
+        }
+    }
+
+    /** Write the contents of this SCollection as a Sparkey file using the default uri. */
+    def asSparkey: SCollection[SparkeyUri] = {
+      val uid = UUID.randomUUID()
+      val uri = ScioUtil.tempLocation(self.context.options) + s"/sparkey-$uid"
```

Comment:
nitpick, you could add log info here, in case someone wants to connect job execution to sparkey key location.